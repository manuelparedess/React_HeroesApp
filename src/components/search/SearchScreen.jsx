import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm"
import { getHeroByName } from "../../selectors/getHeroByName";
import { HeroCard } from "../hero/HeroCard";
import queryString from 'query-string'
import { useMemo } from "react";


export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);

    const [values, handleInputChange] = useForm({
        searchText: q
    });

    const { searchText } = values;

    const heroesFilter = useMemo(() => getHeroByName(q), [q]);


    const handleSearch = (e) => {
        e.preventDefault();

        console.log(searchText)

        navigate(`?q=${searchText}`)
    }


    return (
        <>
            <h1 className="search">Search</h1>
            <hr />

            <div className="row mb-5">

                <div className="col-5">
                    <h4>Search</h4>
                    <hr />

                    <form onSubmit={handleSearch} aria-label='form'>
                        <input
                            type='text'
                            placeholder='Search a hero'
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={searchText}
                            onChange={handleInputChange}
                        />

                        <button
                            type="submit"
                            className="btn btn-outline-primary mt-3"
                        >
                            Searching...
                        </button>
                    </form>

                </div>

                <div className="col-7">

                    <h4>Results</h4>
                    <hr />

                    <div className="row row-cols-1 g-3 searchlist">

                        {
                            (q === '')
                                ? <div className="alert alert-info">Look for a hero</div>
                                : (heroesFilter.length === 0)
                                && <div className="alert alert-danger">No hero with: <b>{q}</b></div>
                        }

                        {

                            heroesFilter.map(hero => (
                                <HeroCard
                                    key={hero.id}
                                    {...hero}
                                />
                            ))
                        }

                    </div>

                </div>

            </div>
        </>
    )
}