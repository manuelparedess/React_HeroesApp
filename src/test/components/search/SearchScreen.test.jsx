import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen'


const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate, /* para solo hacer la simulacion del useNavigate */
}));



describe('Pruebas en SearchScreen', () => {  
    
    test('debe de mostrarse correctamente con los valores por defecto', () => {  

        const {container} = render(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
        expect(screen.getByText('Look for a hero')).toBeTruthy();

    })

    test('debe de mostrar a Batman y el input con el valor del queryString', () => {  

        render(
            <MemoryRouter initialEntries={ ['/search?q=batman'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox')

        expect( input.value ).toBe('batman');

    })

    test('debe de mostrar un error si no se encuentra el heroe', () => {  

        render(
            <MemoryRouter initialEntries={ ['/search?q=batman123'] }>
                <SearchScreen />
            </MemoryRouter>
        );

        expect( screen.getByText('No hero with:') ).toBeTruthy();

    })

    test('debe de llamar el navigate a la nueva pantalla', () => {  

        render(
            <MemoryRouter initialEntries={ ['/search'] }>
                <SearchScreen />
            </MemoryRouter>
        );
        
        const input = screen.getByRole('textbox');
        const form = screen.getByRole('form');

        fireEvent.input( input, {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        fireEvent.submit( form, { preventDefault: () => {} } );

        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
    })

})