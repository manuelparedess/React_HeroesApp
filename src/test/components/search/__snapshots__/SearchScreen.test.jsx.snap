// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchScreen debe de mostrarse correctamente con los valores por defecto 1`] = `
<div>
  <h1
    class="search"
  >
    Search
  </h1>
  <hr />
  <div
    class="row mb-5"
  >
    <div
      class="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-3"
          type="submit"
        >
          Searching...
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="row row-cols-1 g-3 searchlist"
      >
        <div
          class="alert alert-info"
        >
          Look for a hero
        </div>
      </div>
    </div>
  </div>
</div>
`;
